arch/arm/arm64/tls.c

```diff
@@ -0,0 +1,281 @@
+
+extern char _tls_start[], _etdata[], _tls_end[];
+
+/*
+ * This file implements a static exec TLS layout
+ * (variant 2 with one static TLS block).
+ * The calls create a TLS containing space for a TCB.
+ *
+ *             +-------------------------+ \
+ *             | / PADDING 1 / / / / / / | |
+ *    tcbp --> +-------------------------+ |  \
+ *             |                         | |  |
+ *             | Custom TCB format       | |   > Thread Control Block (TCB)
+ *             | (might be used          | |  |  (length: TCB_SIZE)
+ *             |  by a libC)             | |  |
+ *       tlsp -+-> AARCH64_RESERVED      | |  |
+ *          |  +-------------------------+ |  /
+ * data_off<   | / PADDING 2 / / / / / / |  > tls_area_size()
+ *          \->+-------------------------+ |  \
+ *             |                         | |  |
+ *             | .tdata                  | |  |
+ *             |                         | |   > Static TLS block
+ *             + - - - - - - - - - - - - + |  |
+ *             |                         | |  |
+ *             | .tbss                   | |  |
+ *             |                         | |  |
+ *             +-------------------------+ /  /
+ *
+ * Note: ukarch_tlsp_get() points to the last 16 bytes of TCB of the thread.
+ *       AARCH64 ABI requieres the tlsp to point to 2 pointers (16 bytes),
+ *       the pointer to the DTV and the an implementation specific pointer.
+ *
+ *       In reality, this 16 byte area is never used by the compiler
+ *       generated code, so eg. musl just uses it for the last
+ *       member(s) of the TCB.
+ *       (source: https://www.openwall.com/lists/musl/2018/02/09/7
+ *        "Re: TLS storage offsets for TLS_ABOVE_TP", Fri, 9 Feb 2018)
+ */
+
+/*
+ * this is usually read from the elf header
+ * (p_align of the TLS section)
+ *
+ * this NEEDS to match, as aarch64 uses
+ * this at link time to calculate the offset
+ * from the tlsp
+ */
+#define TLS_AREA_ALIGN 16
+
+static __sz tls_data_offset(void)
+{
+	/*
+	 * the distance between tlsp and the start of
+	 * the initial TLS segment is max(16, p_align)
+	 * (source:
+	 * https://fuchsia.dev/fuchsia-src/development/kernel/threads/tls)
+	 */
+	return (AARCH64_RESERVED > TLS_AREA_ALIGN)
+		? AARCH64_RESERVED
+		: TLS_AREA_ALIGN;
+}
+
+static __sz tls_padding_2_size(void)
+{
+	/*
+	 * The tlsp points to the mandatore 16 byte block,
+	 * padding 2 fills the rest of it.
+	 *
+	 * The size of this padding is therefore determined
+	 * by the ABI.
+	 */
+	return tls_data_offset() - AARCH64_RESERVED;
+}
+
+static __sz tls_tdata_size(void)
+{
+	return (__uptr) _etdata  - (__uptr) _tls_start;
+}
+static __sz tls_tbss_size(void)
+{
+	return (__uptr) _tls_end - (__uptr) _etdata;
+}
+
+static __sz tls_padding_1_size(void)
+{
+	/* We do not control:
+	 * - TCB_SIZE
+	 * - TLS_AREA_ALIGN
+	 * - and therefore size of padding 2
+	 * but adding data_offset don't break alignment,
+	 * so we can just align tlsp.
```

Comment:
```suggestion
	 * but adding data_offset doesn't break alignment, so we can just align tlsp.
```
What is `data_offset`?

Reply:
It is `tls_data_offset`, which is also in the shown in the diagram above.

Reply:
Updated